network/test/middleware_test.go

```diff
@@ -268,6 +268,45 @@ func (m *MiddlewareTestSuite) TestEcho() {
 	}
 }
 
+// TestSpoofedPubSubHello evaluates checking the originID of the message w.r.t. its libp2p network ID on PubSub
+// we check a pubsub message with a spoofed OriginID does not get delivered
+// This would be doubled with cryptographic verification of the libp2p network ID in production (see message signing options in pubSub initialization)
+func (m *MiddlewareTestSuite) TestSpoofedPubSubHello() {
+	first := 0
+	last := m.size - 1
+	firstNode := m.ids[first].NodeID
+	lastNode := m.ids[last].NodeID
+
+	// initially subscribe the nodes to the channel
+	for _, mw := range m.mws {
+		err := mw.Subscribe(testChannel)
+		require.NoError(m.Suite.T(), err)
+	}
+
+	// set up waiting for m.size pubsub tags indicating a mesh has formed
+	for i := 0; i < m.size; i++ {
+		select {
+		case <-m.obs:
+		case <-time.After(2 * time.Second):
+			assert.FailNow(m.T(), "could not receive pubsub tag indicating mesh formed")
+		}
+	}
```

Comment:
Could you explain when a pubsub tag is emmitted? Thanks

Reply:
Andy why getting `m.size` of them indicates that a mesh has formed?

Reply:
Libp2p emits `Protect` calls with tag arguments marked by their pubsub topic when establishing each peering in the mesh:
https://github.com/libp2p/go-libp2p-pubsub/blob/master/tag_tracer.go

You would usually wait for [`libp2p.GossipSubD`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L33) such links, but here there's only two nodes.

Reply:
I'd like to understand this a bit more.

From what I understand, for each connection *both* peers would get a `Protect` call for each peering, right?

So if there are 5 peers in total, then we should expect to see 8 tags before we can be confident that a mesh has formed, correct? (5 peers, 5 - 1 = 4 edges required for a connected graph, 4 * 2 = 8 tags).

Here, you are only checking for 5 tags. 

Maybe I'm not understanding how this works?

Reply:
Also, are we sure that `Protect` is only called once for each peer when establishing a peering? In otherwise, is counting the number of tags enough, or would it make sense to actually keep track of the peer IDs in case we get multiple tags for the same peer?

Reply:
So [here's where the calls to `Protect` are intercepted](https://github.com/onflow/flow-go/blob/73d6864593dfc899cd36ef178fcce7f0fad511d6/network/test/testUtil.go#L45-L90). 

[This](https://github.com/libp2p/go-libp2p-pubsub/blob/65f48fb2c8b37563525e3d44042ea2418650f854/trace.go#L468-L493) is where they originate at the libp2p level, in reaction to a join.

I don't think you're incorrect in your analysis. However:

- when I said "You would usually wait for libp2p.GossipSubD such links" I misremembered, we actually wait for  `pubsub.GossipSubD*count` [in tests with count = 10](https://github.com/onflow/flow-go/blob/master/network/test/meshengine_test.go#L169)
- the way I got to that number is by running the test many times at a log level that allowed me to witness the number of tags emitted, 
- indeed, the value there is meant as a fast & loose heuristic to make sure we wait until after the `flow/networking` level that triggers the `UpdatePeers` and then the `libp2p` mesh update downstream of that, NOT as a countdown to full mesh,
- it was nowhere near the theoretical number we would obtain: if, when a node joins a topic, we get 2m graft messages when there are k nodes already on it (m = k if k < D, m = D otherwise, since we don't count pruning), then for 10 nodes we'd expect 90 messages, and we only witness around 65-ish,
- we don't need to finish mesh formation to ensure message delivery, as you noticed we only need a connected graph.

I'm happy to update to a value you'd find more suitable, but for count = 2 (in the test we're commenting on), I note we have, practically and theoretically 2 graft messages, every time.

> are we sure that Protect is only called once for each peer when establishing a peering? 

Yep, the Protect calls otherwise made by our engines are all tagless.

Reply:
Thanks for the explanation :)